RecoTauTag/RecoTau/plugins/PATTauDiscriminantCutMultiplexer.cc

```diff
@@ -0,0 +1,287 @@
+
+    ~PATTauDiscriminantCutMultiplexer();
+    double discriminate(const pat::TauRef&) const override;
+    void beginEvent(const edm::Event& event, const edm::EventSetup& eventSetup) override;
+	
+  private:
+    std::string moduleLabel_;
+	
+    bool loadMVAfromDB_;
+    edm::FileInPath inputFileName_;
+	
+    struct DiscriminantCutEntry
+    {
+    DiscriminantCutEntry()
+      : cutVariable_(),
+        cutFunction_(),
+	mode_(kUndefined)
+    {}
+    ~DiscriminantCutEntry()
+    {
+    }
+    double cutValue_;
+    std::string cutName_;
+    std::unique_ptr<StringObjectFunction<pat::Tau>> cutVariable_;
+    std::unique_ptr<const TGraph> cutFunction_;
+    enum { kUndefined, kFixedCut, kVariableCut };
+    int mode_;
+    };
+    typedef std::map<int, std::unique_ptr<DiscriminantCutEntry>> DiscriminantCutMap;
+    DiscriminantCutMap cuts_;
+	
+    std::string mvaOutputNormalizationName_;
+    std::unique_ptr<const TFormula> mvaOutput_normalization_;
+	
+    bool isInitialized_;
+	
+    edm::InputTag toMultiplex_;
+    edm::InputTag key_;
+    edm::Handle<pat::PATTauDiscriminator> toMultiplexHandle_;
+    edm::Handle<pat::PATTauDiscriminator> keyHandle_;
+    edm::EDGetTokenT<pat::PATTauDiscriminator> toMultiplex_token;
+    edm::EDGetTokenT<pat::PATTauDiscriminator> key_token;
+	
+    int verbosity_;
+};
+
+namespace
+{
+  std::unique_ptr<TFile> openInputFile(const edm::FileInPath& inputFileName) {
+    if ( inputFileName.location() == edm::FileInPath::Unknown){  throw cms::Exception("PATTauDiscriminantCutMultiplexer::loadObjectFromFile") 
+      << " Failed to find File = " << inputFileName << " !!\n";
+    }
+    return std::unique_ptr<TFile>{ new TFile(inputFileName.fullPath().data()) };
+  }
+	
+  template <typename T>
+  std::unique_ptr<const T> loadObjectFromFile(TFile& inputFile, const std::string& objectName)
+  {
+    const T* object = dynamic_cast<T*>(inputFile.Get(objectName.data()));
+    if ( !object )
+      throw cms::Exception("PATTauDiscriminantCutMultiplexer::loadObjectFromFile") 
+	<< " Failed to load Object = " << objectName.data() << " from file = " << inputFile.GetName() << " !!\n";
+    //Need to use TObject::Clone since the type T might be a base class
+    std::unique_ptr<const T> copy{ static_cast<T*>(object->Clone()) };
+	
+    return std::move(copy);
+  }
+	
+  std::unique_ptr<const TGraph> loadTGraphFromDB(const edm::EventSetup& es, const std::string& graphName, const int& verbosity_ = 0)
+  {
+    if(verbosity_){
+      std::cout << "<loadTGraphFromDB>:" << std::endl;
+      std::cout << " graphName = " << graphName << std::endl;
+    }
+    edm::ESHandle<PhysicsTGraphPayload> graphPayload;
+    es.get<PhysicsTGraphPayloadRcd>().get(graphName, graphPayload);
+    return std::unique_ptr<const TGraph>{ new TGraph(*graphPayload.product()) };
+  }  
+	
+  std::unique_ptr<TFormula> loadTFormulaFromDB(const edm::EventSetup& es, const std::string& formulaName, const TString& newName, const int& verbosity_ = 0)
```

Comment:
When switching to FormulaEvaluator, please check that the expressions you pass get the same answers as when you use TFormula. We've had a few cases where they were not the same. When that happens, we've been able to find an alternate way to express the same thing that works and then fix the bug that caused the expression to fail.